spec/mongoid/search_indexable_spec.rb

```diff
@@ -0,0 +1,139 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+class SearchIndexHelper
+  attr_reader :model
+
+  def initialize(model)
+    @model = model
+    model.collection.create
+  end
+
+  def collection
+    model.collection
+  end
+
+  # Wait for all of the indexes with the given names to be ready; then return
+  # the list of index definitions corresponding to those names.
+  def wait_for(*names, &condition)
+    names.flatten!
+
+    timeboxed_wait do
+      result = collection.search_indexes
+      return filter_results(result, names) if names.all? { |name| ready?(result, name, &condition) }
+    end
+  end
+
+  # Wait until all of the indexes with the given names are absent from the
+  # search index list.
+  def wait_for_absense_of(*names)
+    names.flatten.each do |name|
+      timeboxed_wait do
+        break if collection.search_indexes(name: name).empty?
+      end
+    end
+  end
+
+  private
+
+  def timeboxed_wait(step: 5, max: 300)
+    start = Mongo::Utils.monotonic_time
+
+    loop do
+      yield
+
+      sleep step
+      raise Timeout::Error, 'wait took too long' if Mongo::Utils.monotonic_time - start > max
```

Comment:
A few points re: timeout.

- Large collections (100M+ docs) should be considered the _canonical use case_ for search indexes.
- First, indexing large collections is going to take much longer than 5 minutes. I would propose there should be _no timeout_, and instead, indicate the index building progress. (Atlas UI does this--you can see the % completion of index--so there is presumably some way Mongoid could as well.
   - Consider that is it is frustrating to make a user wait 5 minutes then fail :)
- Second, if you set socket_timeout in your Mongoid.yml--which also is a best practice--your index building here is going to timeout after 30 seconds. Index building should skip socket timeout entirely, since it happens in background anyway.
- Finally, whatever logic you add here should be backported to traditional index building as well, and possible extracted out to a generic class which applies to both.